test/kafka-upsert-benchmark/setup.td

```diff
@@ -0,0 +1,38 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+#
+# Setup for performing load tests for persistent upsert sources.
+#
+
+$ kafka-create-topic topic=load-test partitions=4
+
+> CREATE SOURCE load_test
+  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-load-test-${testdrive.seed}'
+  FORMAT BYTES
+  INCLUDE PARTITION AS kafka_partition, OFFSET AS mz_offset
+  ENVELOPE UPSERT
+
+# Keep track of the maximum offset afor each partition that has been ingested and
+# made definite, and try do so with as little overhead as possible.
+> CREATE MATERIALIZED VIEW max_offsets AS SELECT
+  kafka_partition,
+  MAX(mz_offset) as max_offset FROM load_test
```

Comment:
I think `MAX` should be considered as a "big overhead" operation, given that it will keep a list of all the mz_offsets ever seen in memory.

It may be best to simply use `COUNT(*)` instead.

Reply:
Right. I considered this in depth because count(*) is more memory efficient but doesn't offer a good way to track when we've caught up when there are duplicates (because a duplicate key won't show up in the counts twice)

Reply:
I came up with a solution that I like (basically, we track an approximate max with `mz_offset / 1000 + 1` aka the mz_offset to the next thousandth offset, which lets us use a max with 0.1% of the memory overhead. The math around tracking "have we caught up" gets more complicated so I'm going to do that in a followup

Reply:
> Right. I considered this in depth because count(*) is more memory efficient but doesn't offer a good way to track when we've caught up when there are duplicates (because a duplicate key won't show up in the counts twice)

The trick I used with the feature benchmark framework is to emit a unique marker record that has not been generated  before. Once I see this record in the output, I know that we have caught up to that point, regardless of any duplicates and such that may have been present in the stream prior to that.

Reply:
sorry could you say a bit more about what the feature benchmarks are computing? (do they just `CREATE MATERIALIZED SOURCE ...` or do they do something else?)

Reply:
I was talking about this scenario:

https://github.com/MaterializeInc/materialize/blob/main/test/feature-benchmark/scenarios.py#L716

Basically we ingest a bunch of `key=1`'s followed by one `key=2` . Then we measure the time it takes from `CREATE MATERIALIZED SOURCE` to the time it takes for the `2` to show up in the output of the `SELECT`.

Reply:
I actually came up with another approach that I think works even better.

Basically, you can use `mz_source_info` but keep a wholly uncompacted version around, and then you just need to:

1. look up the current time the dataflow you are testing is materialized up to
2. look up which how much data was recorded in mz_source_info AS OF that time,

this cuts the memory requirement down by a lot because now the dataflow for the source can be a simple count(*) (according to docker stats on my laptop a test sending 10k records per second for ~120 seconds goes from ~850 MB to ~350 MB)